resize.js

```diff
@@ -15,218 +18,216 @@ var S3 = new AWS.S3();
-    sentToken = req.headers['x-token']
-    consumeToken.run([sentToken], function(err) {
-      if ( !err && this.changes === 1 ) {
-        // If there is not error and the token was valid
-        matches = req.url.match(/^\/(.*)\.([^.]+)$/);
-        if ( supportedFileType(matches[2]) ) {
-            // And we support the filetype
-            log('info','Starting to write original file ' + matches[1]);
-            original = fs.createWriteStream(config.get('originals_dir') + '/' + matches[1]);
-            r = req.pipe(original);
-            r.on('finish', function() {
-              log('info','Finished writing original file ' + matches[1]);
-              res.writeHead(200, {'Content-Type': 'application/json'});
-              res.write(JSON.stringify({'status': 'OK', 'id': matches[1]}));
-              res.end();
-           });
+    gm(file)
+        .options({imageMagick: true})
+        .resize(resolutionX, resolutionY)
+        .toBuffer(fileType, function (err, stream) {
+            if (!err) {
+                // This might mean we have generated the same file while an upload was in progress.
+                // However this is still better than not being able to server the image
+                Image.uploadToCache(fileName, fileType, resolutionX, resolutionY, stream);
+            }
+        });
+};
+Image.uploadToCache = function (fileName, fileType, resolutionX, resolutionY, content) {
+    // Upload to AWS
+    var key = fileName + '_' + resolutionX + 'x' + resolutionY + '.' + fileType;
+    var upload_params = {
+        Bucket: config.get('aws.bucket'),
+        Key: key,
+        ACL: 'public-read',
+        Body: content,
+        // We let the client cache this for a month
+        Expires: (new Date()).setMonth(new Date().getMonth() + 1) / 1000,
+        ContentType: supportedFileType(fileType),
+        // We let any intermediate server cache this result as well
+        CacheControl: 'public'
+    };
+    S3.putObject(upload_params, function (err, data) {
+        if (err) {
+            log('error', 'AWS upload error: ' + JSON.stringify(err));
         } else {
-          res.writeHead(415, 'Image type not supported');
-          res.end();
+            log('info', 'Uploading of ' + key + ' went very well');
+            var url = config.get('aws.bucket_url') + '/' + key;
+            insertImage.run([fileName, resolutionX, resolutionY, supportedFileType(fileType), url], function (err) {
+                if (err) {
+                    console.error(err);
+                }
+            });
         }
-      } else {
-        log('warn','Invalid or expired token used for upload');
-        res.writeHead('403','Forbidden');
+    });
+};
+Image.upload = function (req, res) {
+    // Upload the RAW image to AWS S3, stripped of its extension
+    // First check the token
+    var sentToken = req.headers['x-token'];
+    var matches = req.url.match(/^\/(.*)\.([^.]+)$/);
+    if (supportedFileType(matches[2])) {
```

Comment:
This is probably a security vulnerability: you match on file extension, but that can be anything. Without extension S3 won't set any content-type headers and your browser will instead interpret the content-type based on it's content. So if you just check on extension here, then the user can upload any file as if it were an image, and the browser will happily download it and interpret it for it's actual file type.

So either you'll need to keep the file extension such that S3 will send the correct headers, or you'll need to do some mime-type sniffing to block non-image files.